cloud/pkg/cloudhub/handler/messagehandler.go

```diff
@@ -300,59 +315,210 @@ func (mh *MessageHandle) GetNodeCount() int {
+		return
+	}
+
 	for {
-		msg, err := messages.Get()
-		if err != nil {
-			klog.Errorf("failed to consume event for node %s, reason: %s", info.NodeID, err.Error())
-			if err.Error() == MsgFormatError {
-				// error format message should not impact other message
-				messages.Ack()
-				continue
-			}
-			stop <- messageQueueDisconnect
+		select {
+		case <-stopSendMsg:
+			klog.Errorf("Node %s disconnected and stopped sending messages", info.NodeID)
 			return
+		default:
+			mh.handleMessage(nodeQueue, nodeStore, hi, info, stopServe, "message")
 		}
+	}
+}
 
-		if model.IsNodeStopped(msg) {
-			klog.Infof("node %s is stopped, will disconnect", info.NodeID)
-			messages.Ack()
-			stop <- nodeStop
-			return
-		}
-		if !model.IsToEdge(msg) {
-			klog.Infof("skip only to cloud event for node %s, %s, content %s", info.NodeID, dumpMessageMetadata(msg), msg.Content)
-			messages.Ack()
-			continue
+func (mh *MessageHandle) handleMessage(nodeQueue workqueue.RateLimitingInterface,
+	nodeStore cache.Store, hi hubio.CloudHubIO,
+	info *model.HubInfo, stopServe chan ExitCode, msgType string) {
+	key, quit := nodeQueue.Get()
+	if quit {
+		klog.Errorf("nodeQueue for node %s has shutdown", info.NodeID)
+		return
+	}
+	obj, exist, _ := nodeStore.GetByKey(key.(string))
+	if !exist {
+		klog.Errorf("nodeStore for node %s doesn't exist", info.NodeID)
+		return
+	}
+
+	msg := obj.(*beehiveModel.Message)
+
+	if model.IsNodeStopped(msg) {
+		klog.Infof("node %s is stopped, will disconnect", info.NodeID)
+		stopServe <- nodeStop
+		return
+	}
+	if !model.IsToEdge(msg) {
+		klog.Infof("skip only to cloud event for node %s, %s, content %s", info.NodeID, dumpMessageMetadata(msg), msg.Content)
+		return
+	}
+	klog.V(4).Infof("event to send for node %s, %s, content %s", info.NodeID, dumpMessageMetadata(msg), msg.Content)
+
+	copyMsg := deepcopy(msg)
+	trimMessage(msg)
+	err := hi.SetWriteDeadline(time.Now().Add(time.Duration(mh.WriteTimeout) * time.Second))
+	if err != nil {
+		klog.Errorf("SetWriteDeadline error, %s", err.Error())
+		stopServe <- hubioWriteFail
+		return
+	}
+	if msgType == "listMessage" {
+		mh.send(hi, info, msg)
+		// delete successfully sent events from the queue/store
+		nodeStore.Delete(msg)
+	} else {
+		mh.sendMsg(hi, info, msg, copyMsg, nodeStore)
+	}
+
+	nodeQueue.Done(key)
+}
+
+func (mh *MessageHandle) sendMsg(hi hubio.CloudHubIO, info *model.HubInfo, msg, copyMsg *beehiveModel.Message, nodeStore cache.Store) {
+	mh.MessageAckChannel[msg.GetID()] = make(chan struct{})
+
+	// initialize timer and retry count for sending message
+	var (
+		retry                       = 0
+		retryInterval time.Duration = 5
+	)
+	ticker := time.NewTimer(retryInterval * time.Second)
+	mh.send(hi, info, msg)
+
+LOOP:
+	for {
+		select {
+		case <-mh.MessageAckChannel[msg.GetID()]:
+			delete(mh.MessageAckChannel, msg.GetID())
+			mh.saveSuccessPoint(copyMsg, info, nodeStore)
+			break LOOP
+		case <-ticker.C:
+			if retry == 4 {
+				break LOOP
```

Comment:
After the number of failed messages, do we need to rejoin the queue and resend regularly？

Reply:
Let's add it to the track issue first.